File: ext/src/Cassandra/Table.c
/**
 * Copyright 2015-2016 DataStax, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "php_cassandra.h"

zend_class_entry *cassandra_table_ce = NULL;

ZEND_BEGIN_ARG_INFO_EX(arginfo_name, 0, ZEND_RETURN_VALUE, 1)
  ZEND_ARG_INFO(0, name)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_none, 0, ZEND_RETURN_VALUE, 0)
ZEND_END_ARG_INFO()

static zend_function_entry cassandra_table_methods[] = {
  PHP_ABSTRACT_ME(Table, name, arginfo_none)
  PHP_ABSTRACT_ME(Table, comment, arginfo_none)
  PHP_ABSTRACT_ME(Table, readRepairChance, arginfo_none)
  PHP_ABSTRACT_ME(Table, localReadRepairChance, arginfo_none)
  PHP_ABSTRACT_ME(Table, gcGraceSeconds, arginfo_none)
  PHP_ABSTRACT_ME(Table, caching, arginfo_none)
  PHP_ABSTRACT_ME(Table, bloomFilterFPChance, arginfo_none)
  PHP_ABSTRACT_ME(Table, memtableFlushPeriodMs, arginfo_none)
  PHP_ABSTRACT_ME(Table, defaultTTL, arginfo_none)
  PHP_ABSTRACT_ME(Table, speculativeRetry, arginfo_none)
  PHP_ABSTRACT_ME(Table, indexInterval, arginfo_none)
  PHP_ABSTRACT_ME(Table, compactionStrategyClassName, arginfo_none)
  PHP_ABSTRACT_ME(Table, compactionStrategyOptions, arginfo_none)
  PHP_ABSTRACT_ME(Table, compressionParameters, arginfo_none)
  PHP_ABSTRACT_ME(Table, populateIOCacheOnFlush, arginfo_none)
  PHP_ABSTRACT_ME(Table, replicateOnWrite, arginfo_none)
  PHP_ABSTRACT_ME(Table, maxIndexInterval, arginfo_none)
  PHP_ABSTRACT_ME(Table, minIndexInterval, arginfo_none)
  PHP_ABSTRACT_ME(Table, column, arginfo_name)
  PHP_ABSTRACT_ME(Table, columns, arginfo_none)
  PHP_FE_END
};

void cassandra_define_Table(TSRMLS_D)
{
  zend_class_entry ce;

  INIT_CLASS_ENTRY(ce, "Cassandra\\Table", cassandra_table_methods);
  cassandra_table_ce = zend_register_internal_class(&ce TSRMLS_CC);
  cassandra_table_ce->ce_flags |= ZEND_ACC_INTERFACE;
}

File: docs/Cassandra/Table.php
<?php

/**
 * Copyright 2017 DataStax, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Cassandra;

/**
 * A PHP representation of a table
 */
interface Table {

    /**
     * Returns the name of this table
     *
     * @return string Name of the table
     */
    public function name();

    /**
     * Return a table's option by name
     *
     * @param string $name The name of the option
     *
     * @return \Cassandra\Value Value of an option by name
     */
    public function option($name);

    /**
     * Returns all the table's options
     *
     * @return array A dictionary of `string` and `Value` pairs of the table's options.
     */
    public function options();

    /**
     * Description of the table, if any
     *
     * @return string Table description or null
     */
    public function comment();

    /**
     * Returns read repair chance
     *
     * @return float Read repair chance
     */
    public function readRepairChance();

    /**
     * Returns local read repair chance
     *
     * @return float Local read repair chance
     */
    public function localReadRepairChance();

    /**
     * Returns GC grace seconds
     *
     * @return int GC grace seconds
     */
    public function gcGraceSeconds();

    /**
     * Returns caching options
     *
     * @return string Caching options
     */
    public function caching();

    /**
     * Returns bloom filter FP chance
     *
     * @return float Bloom filter FP chance
     */
    public function bloomFilterFPChance();

    /**
     * Returns memtable flush period in milliseconds
     *
     * @return int Memtable flush period in milliseconds
     */
    public function memtableFlushPeriodMs();

    /**
     * Returns default TTL.
     *
     * @return int Default TTL.
     */
    public function defaultTTL();

    /**
     * Returns speculative retry.
     *
     * @return string Speculative retry.
     */
    public function speculativeRetry();

    /**
     * Returns index interval
     *
     * @return int Index interval
     */
    public function indexInterval();

    /**
     * Returns compaction strategy class name
     *
     * @return string Compaction strategy class name
     */
    public function compactionStrategyClassName();

    /**
     * Returns compaction strategy options
     *
     * @return \Cassandra\Map Compaction strategy options
     */
    public function compactionStrategyOptions();

    /**
     * Returns compression parameters
     *
     * @return \Cassandra\Map Compression parameters
     */
    public function compressionParameters();

    /**
     * Returns whether or not the `populate_io_cache_on_flush` is true
     *
     * @return bool Value of `populate_io_cache_on_flush` or null
     */
    public function populateIOCacheOnFlush();

    /**
     * Returns whether or not the `replicate_on_write` is true
     *
     * @return bool Value of `replicate_on_write` or null
     */
    public function replicateOnWrite();

    /**
     * Returns the value of `max_index_interval`
     *
     * @return int Value of `max_index_interval` or null
     */
    public function maxIndexInterval();

    /**
     * Returns the value of `min_index_interval`
     *
     * @return int Value of `min_index_interval` or null
     */
    public function minIndexInterval();

    /**
     * Returns column by name
     *
     * @param string $name Name of the column
     *
     * @return \Cassandra\Column Column instance
     */
    public function column($name);

    /**
     * Returns all columns in this table
     *
     * @return array A list of Column instances
     */
    public function columns();

    /**
     * Returns the partition key columns of the table
     *
     * @return array A list of of Column instances
     */
    public function partitionKey();

    /**
     * Returns both the partition and clustering key columns of the table
     *
     * @return array A list of of Column instances
     */
    public function primaryKey();

    /**
     * Returns the clustering key columns of the table
     *
     * @return array A list of of Column instances
     */
    public function clusteringKey();

    /**
     * @return array A list of cluster column orders ('asc' and 'desc')
     */
    public function clusteringOrder();

}
